### ISSUE TITLE
Shade or otherwise fix ASM 4.x and 3.x conflict in 0.5.1

### ISSUE THREAD

**Issue by dalaro** [thinkaurelius/titan#784]
titan-solr brought ASM 4.1 back into the Titan dependency tree.  This is going to be more delicate to resolve than with kryo-shaded, since the solr-core artifact on Maven Central does not use a shaded ASM.  Ideally we want a single fix that works not just in the zipfile but also interops with the solr-core artifact on Central.  Not sure that's even possible without putting an onerous runtime environment requirement on the user (container).

```
Exception in thread "main" java.lang.VerifyError: class org.codehaus.groovy.runtime.callsite.GroovySunClassLoader$2 overrides final method visit.(IILjava/lang/String;Ljava/lang/String;Ljava/lang/String;[Ljava/lang/String;)V
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:455)
        at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:367)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at org.codehaus.groovy.runtime.callsite.CallSiteGenerator.isCompilable(CallSiteGenerator.java:245)
        at org.codehaus.groovy.reflection.CachedMethod.createStaticMetaMethodSite(CachedMethod.java:284)
        at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite.createStaticMetaMethodSite(StaticMetaMethodSite.java:111)
        at groovy.lang.MetaClassImpl.createStaticSite(MetaClassImpl.java:3034)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallStaticSite(CallSiteArray.java:61)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallSite(CallSiteArray.java:146)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:42)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:112)
        at org.codehaus.groovy.tools.shell.Groovysh.<clinit>(Groovysh.groovy:40)
        at com.thinkaurelius.titan.hadoop.tinkerpop.gremlin.Console.<init>(Console.java:37)
        at com.thinkaurelius.titan.hadoop.tinkerpop.gremlin.Console.<init>(Console.java:70)
        at com.thinkaurelius.titan.hadoop.tinkerpop.gremlin.Console.main(Console.java:75)
```

**Comment by dalaro** [comment]
Related discussion: https://groups.google.com/d/msg/aureliusgraphs/MSC779hz-uA/jk5hKPhcUc4J

**Comment by dalaro** [comment]
Previous incarnation of this issue: #361

**Comment by dalaro** [comment]
It might be easiest to leave the 4.x dependency dragged in by solr-core unshaded, and then attempt to shade the 3.x dependencies instead.  The path to 3.x through hbase-server can probably be ignored, since it's just used in testing.

**Comment by dalaro** [comment]
I implemented the strategy suggested in my last comment.  titan05 now depends only on ASM 4.1 via Solr.  The other copies of ASM are all shaded.  These changes were tricky and could use some additional testing, but I think this issue is fixed (until we add another dependency that unexpectedly drags ASM back into Titan's tree).

Several projects upstream independently decided to shade ASM themselves (jersey-server, avro, and groovy).  Newer versions of these artifacts dropped the explicit ASM dependency in favor of shipping a shaded copy in the artifact itself.  Updating to newer versions of these three artifacts let me cut all the dependency chains leading to ASM 3.x in the current titan05 codebase.  I normally wouldn't update dependency versions in the middle of a bugfix cycle, but this is probably the least worst approach for continued Solr integration.

The one that makes me nervous is groovy.  We jumped from 1.8.9 (the version against which TP 2.5.0 was compiled) to 2.3.7.  There is no 1.x.y release after 1.8.9 to which we could have upgraded; looks like 2.x.y or bust.

The only bits I had to recompile were inside Titan's Console.java, where we configure a JLine history object.  Groovy 2.3.7 uses JLine 2.x, which introduced backwards-incompatible changes to the command history package versus 1.8.9's old JLine 0.9/1.x.  I ran the titan-dist integration tests, ran the gremlin 2.5.0 test suite with Groovy 2.3.7, and played a bit with gremlin.sh manually.  All these tests checked out, but this is still a major change and hard to verify.

Aside: After these changes, I looped over Titan's set of jars, listed the contents of each, and grepped for `asm`.  This is a useful complement to `mvn dependency:tree` since it works on the classfile level instead of the Maven metadata level.  I found seven distinct copies of shaded ASM packages in use by the following packages:
- BDB JE
- Elasticsearch
- Glassfish
- Groovy 
- Guice
- Jersey
- Kryo

**Comment by dalaro** [comment]
Dropping in Groovy 2.3.7 at runtime breaks Rexster's console through the JLine upgrade:

```
Exception in thread "main" java.lang.NoClassDefFoundError: jline/ConsoleReader
        at com.tinkerpop.rexster.console.RexsterConsole.getInputReader(RexsterConsole.java:196)
        at com.tinkerpop.rexster.console.RexsterConsole.primaryLoop(RexsterConsole.java:97)
        at com.tinkerpop.rexster.console.RexsterConsole.acceptReplCommands(RexsterConsole.java:71)
        at com.tinkerpop.rexster.console.RexsterConsole.start(RexsterConsole.java:53)
        at com.tinkerpop.rexster.console.RexsterConsole.main(RexsterConsole.java:381)
Caused by: java.lang.ClassNotFoundException: jline.ConsoleReader
        at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
        ... 5 more
```

Although I suspect that it might be possible to resolve this by retaining jline 1.x in a separate directory for use only by rexster-console, I'm reluctant to give up on the unitary classpath we've maintained thus far.

I discussed this with Stephen.  It sounds like a Groovy version bump is very unlikely during TP2's lifetime, so it's also unlikely that Rexster will ever be compatible with JLine 2.x.  I think I'm going to have to resort to shading gremlin-groovy to hide ASM.  This shaded copy of gremlin-groovy will need to be on Central (under some groupId that makes it clear it's been adapted specifically to Titan, not part of Tinkerpop proper).  This would let us retain JLine 0.9/1.x and should not necessarily require a shaded Rexster artifact.

From a user perspective, this is going to turn Titan's gremlin-groovy into (more of?) a "black box", and bind Titan 0.5.x releases tighter to specific TP 2.x.y releases, since upgrading to a newer TP 2.x.y may require building a new shaded gremlin-groovy artifact that hides ASM 3.x to avoid conflicting with the ASM 4.x dependency in Titan's Solr module.

This will not, however, affect Titan 0.9, since TP3 has been Groovy 2.x.y based from the start.

**Comment by dalaro** [comment]
I rolled back to Groovy 1.8.9 in bfcc99dbd71e90f4a2a17b24cf02ab484bf89396.

I wanted to avoid pushing a shaded gremlin-groovy doppelganger artifact to Central, if possible, since that would become obsolete with TP3 anyway, and Central is forever.  I configured titan-core to contain a copy of org.codehaus.groovy:groovy and asm:\* (3.2).  The ASM packages are relocated to com.thinkaurelius.shaded.groovy.asm, and the Groovy classes packed into titan-core that rely on ASM are rewritten to use that relocated package prefix.  I also excluded groovy from titan-core's dependencies, since it now comes with titan-core, and tweaked the titan-dist pom to match.

That commit also introduced a new maven-enforcer-plugin execution using the 'bannedDependencies` rule to prevent ASM 3.x from quietly creeping back into Titan's dependency tree unnoticed.
